src/js/media-controller.js

```diff
@@ -380,40 +418,76 @@ class MediaController extends MediaContainer {
   }
 }
 
-/*
-  Loop through child nodes and set the media[State] on every child.
-*/
-function propagateMediaState(nodeList, stateName, val) {
-  Array.from(nodeList).forEach(child => {
-    // All elements we care about at least have an empty children list (i.e. not <style>)
-    if (!child.children) return;
+const MEDIA_UI_ATTRIBUTE_NAMES = Object.values(MediaUIAttributes);
 
-    const childName = child.nodeName.toLowerCase();
+const getMediaUIAttributesFrom = (child) => {
+  const { constructor: { observedAttributes } } = child;
+  const mediaChromeAttributesList = child?.getAttribute?.(MediaUIAttributes.MEDIA_CHROME_ATTRIBUTES)?.split?.(/\s+/);
+  if (!Array.isArray(observedAttributes || mediaChromeAttributesList)) return [];
+  return (observedAttributes || mediaChromeAttributesList).filter(attrName => MEDIA_UI_ATTRIBUTE_NAMES.includes(attrName));
+}
 
-    // Don't propagate into media elements, UI can't live in <video>
-    // so just avoid potential conflicts
-    if (child.slot === 'media') return;
-
-    function setAndPropagate() {
-      // Only set if previously defined, at least as null
-      // This is how element authors can tell us they want to
-      // receive these state updates
-      if (typeof child[stateName] !== 'undefined') {
-        child[stateName] = val;
-      }
+const isMediaUIElement = (child) => !!getMediaUIAttributesFrom(child).length;
 
-      propagateMediaState(child.children, stateName, val);
+const setAttr = (child, attrName, attrValue) => {
+  if (typeof attrValue === 'boolean') {
+    if (attrValue) return child.setAttribute(attrName, '');
+    return child.removeAttribute(attrName);
+  }
+  return child.setAttribute(attrName, attrValue);
+}
 
-      // We might consider an option to block piercing the shadow dom
-      if (child.shadowRoot) propagateMediaState(child.shadowRoot.childNodes, stateName, val);
+const getMediaUIElementDescendants = (rootNode) => {
+  const { childNodes } = rootNode;
+  const rootMediaUIElements = isMediaUIElement(rootNode) ? [rootNode] : [];
+  /* leaf node that is either a mediaChromeNode or not */
+  if (!(childNodes && childNodes.length)) return rootMediaUIElements;
+  return [
+    ...rootMediaUIElements,
+    ...Array.prototype.flatMap.call(childNodes, getMediaUIElementDescendants)
+  ];
+};
+
+const toNextMediaUIElementsState = (mutationsList = []) => {
```

Comment:
@cjpillsbury I think I'm going to need you to walk me through this a little in person, but there's a few things I want to confirm:
* This isn't only looking at the children of MediaUI nodes, because they can be children of non-mediaUI nodes
* We're not drilling into the children of media elements

Reply:
- awww horesapples it won't work if a child is added with a media ui element descendant in one fell swoop. Will make recursive a la initial dom tree traversal.
- Given what I just mentioned about the above, this won't currently drill down into the children of media ui elements. That said, don't we want it to drill into the children of media ui elements? see, e.g. https://github.com/muxinc/media-chrome/blob/main/src/js/media-time-range.js#L61 (at least the "property-based equivalent" in the old architecture passed those props directly to the `media-thumb-preview` in the `media-time-range`).

Might be worth talking through this one/these broad questions in more detail.

Reply:
Yeah, sorry. Do drill into `<div>`. Do drill into `<media-play-button>` etc. Do not drill into `<something slot=“media”>`.

Reply:
From this conversation I'm not exactly clear on where we landed, but I think we have a bug on current main related to this change.

> Given what I just mentioned about the above, this won't currently drill down into the children of media ui elements.

Based on this comment, it seems like @cjpillsbury is aware, but just wanted to surface what I'm finding.

Here `media-thumbnail-preview` is asking for `time media-preview-image media-preview-coords` (these also exist in the `observedAttributes` on the constructor.

<img width="977" alt="media-thumbnail-preview_2021-08-13_16-06-25" src="https://user-images.githubusercontent.com/764988/129425952-6af5e2fd-1115-4f7b-9089-1f2696a5e68b.png">


Those attributes are not getting updated, and the reason for that that appears to be b/c `media-thumbnail-preview` is a child of `media-time-range` and therefore it's not on the list of `associatedElements`.

When I dig into `this.associatedElements` on the `media-controller` I see it has these 10 elements and `media-thumbnail-preview` is missing from the list:

```
0: media-controller
1: media-seek-forward-button
2: media-fullscreen-button
3: media-mute-button
4: media-pip-button
5: media-play-button
6: media-playback-rate-button
7: media-time-range
8: media-seek-backward-button
9: media-volume-range
```

Reply:
Yup there's a pending PR that will resolve this. W

> I like this rearchitecture a lot
> 
> > media-chrome UI elements will "announce" the media-chrome-attributes they expect
> 
> It looks like two steps are required for this: (1) use `observedAttributes[]` on the constructor and (2) set `media-chrome-attributes=` attribute on the element.
> 
> If I understand this correctly `media-controller` pays attention to `media-chrome-attributes` and propagates state based on that whereas the element itself uses `observedAttributes` in order to get the `attributeChangedCallback` which it will need to update the UI.
> 
> Typically these end up being the same list, but I guess they don't strictly have to be.

Yeah, there's a pending PR to handle this. High level:
- We want to allow several ways to "automagically" associate/unassociate Media UI Elements (e.g. both 1 & 2 you describe)
- But our "built in"/"default" Media UI Elements shouldn't (redundantly) need to declare both.

(PR is currently issued to my repo to be able to see diff, pending a blocking PR merge: https://github.com/cjpillsbury/media-chrome/pull/2/files)